src/Sylius/Channel/ProChannelContext.php

```diff
@@ -0,0 +1,34 @@
+<?php
+declare(strict_types=1);
+
+namespace AppBundle\Sylius\Channel;
+
+use Sylius\Component\Channel\Context\ChannelContextInterface;
+use Sylius\Component\Channel\Model\ChannelInterface;
+use Sylius\Component\Channel\Repository\ChannelRepositoryInterface;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\RequestStack;
+
+class ProChannelContext implements ChannelContextInterface
+{
+    /**
+     * @var ChannelRepositoryInterface
+     */
+    private ChannelRepositoryInterface $channelRepository;
+    /**
+     * @var Request|null
+     */
+    private ?Request $masterRequest;
+
+    public function __construct(ChannelRepositoryInterface $channelRepository, RequestStack $request)
+    {
+
+        $this->channelRepository = $channelRepository;
+        $this->masterRequest = $request->getMasterRequest();
+    }
+
+    public function getChannel(): ChannelInterface
+    {
+        return $this->channelRepository->findOneByCode($this->masterRequest->cookies->get('channel_cart', 'web'));
```

Comment:
As the cookie is written during the `kernel.response` event, `? change_channel=pro` does not have effect immediately, but on the subsequent request. The value in the request parameter should have a priority over the cookie. See how it's done in [FakeChannelCodeProvider](https://github.com/Sylius/Sylius/blob/34fcad9732bb53eab1d02a721900bb2b3e1edda2/src/Sylius/Bundle/ChannelBundle/Context/FakeChannel/FakeChannelCodeProvider.php)

Reply:
This is still a problem @flug